promql/query_logger.go

```diff
@@ -0,0 +1,217 @@
+
+func getMMapedFile(filename string, filesize int, logger log.Logger) (error, []byte) {
+
+	fd, err := os.OpenFile(filename, os.O_CREATE|os.O_RDWR|os.O_TRUNC, 0644)
+	if err != nil {
+		level.Error(logger).Log("msg", "Error opening query log file", "file", filename, "err", err)
+		return err, []byte{}
+	}
+
+	if err := fd.Truncate(int64(filesize)); err != nil {
+		level.Error(logger).Log("msg", "Error creating file of specified size", "file", filename, "Attemped size", filesize, "err", err)
+		return err, []byte{}
+	}
+
+	fileAsBytes, err := syscall.Mmap(int(fd.Fd()), 0, filesize, syscall.PROT_READ|syscall.PROT_WRITE, syscall.MAP_SHARED)
+	if err != nil {
+		level.Error(logger).Log("msg", "Error mmapping file", "file", filename, "err", err)
+		return err, []byte{}
+	}
+
+	return err, fileAsBytes
+}
+
+func NewQueryLogger(filename string, filesize int, logger log.Logger) (queryLogger QueryLogger) {
+	defer func() {
+		if r := recover(); r != nil {
+			queryLogger = QueryLogger{}
+		}
+	}()
+
+	logUnfinishedQueries(filename, filesize, logger)
+
+	if filesize < 1+entrySize {
+		level.Error(logger).Log("msg", "Filesize must be greater than entrysize.", "entrysize", entrySize)
+		return QueryLogger{}
+	}
+
+	err, fileAsBytes := getMMapedFile(filename, filesize, logger)
+	if err != nil {
+		return QueryLogger{}
+	}
+
+	copy(fileAsBytes, "[")
+	queryLogger = QueryLogger{
+		mmapedFile:   fileAsBytes,
+		getNextIndex: make(chan int, (1<<16)-1),
+		lastIndex:    make(chan int),
+		logger:       logger,
+		Ok:           true,
+	}
+
+	go queryLogger.generateLastIndex(filesize)
+
+	return queryLogger
+}
+
+type Entry struct {
+	Query     string `json:"query"`
+	Timestamp int64  `json:"timestamp"`
+}
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func newJsonEntry(query string, timestamp int64, logger log.Logger) []byte {
+	// 35 bytes is the size of json entry for empty string. Including a comma for every entry, max query size is entrySize-36
+	entry := Entry{query[:min(len(query), entrySize-36)], timestamp}
+	jsonEntry, err := json.Marshal(entry)
+
+	if err != nil {
+		level.Error(logger).Log("msg", "Cannot create json of query", "query", query)
+		return []byte{}
+	}
+	return jsonEntry
+}
+
+// Adds an entry struct -- json object for each query indicating its timestamp and name -- to query log file
+func addEntry(file []byte, i int, timestamp int64, query string, logger log.Logger) {
+	entry := newJsonEntry(query, timestamp, logger)
+	start := i
+	blankStart := i + len(entry)
+	end := i + entrySize
+
+	copy(file[start:], entry)
+	copy(file[blankStart:end], strings.Repeat(" ", end-blankStart))
+	copy(file[end-1:], ",")
+}
+
+func (logger QueryLogger) generateLastIndex(maxSize int) {
+	for i := 1; i <= maxSize-entrySize; i += entrySize {
+		logger.lastIndex <- i
+	}
+	close(logger.lastIndex)
+}
+
+func (logger QueryLogger) LogQuery(query string, queryCompleted <-chan struct{}) {
+	defer func() {
+		if r := recover(); r != nil {
```

Comment:
How can this happen?

Reply:
It can't. I kept it as fail safe mechanism to keep prometheus from crashing in case something changes in this code in the future. Should I remove it?

Reply:
Yes, remove it. If there's a bug we'll deal with it when it's discovered.